src/ott/tools/map_estimator.py

```diff
@@ -0,0 +1,264 @@
+# Copyright OTT-JAX
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+from functools import partial
+from typing import Any, Callable, Dict, Iterator, Optional, Tuple
+
+import jax
+import jax.numpy as jnp
+import optax
+from flax.core.frozen_dict import FrozenDict
+from flax.training import train_state
+
+from ott.solvers.nn.models import ModelBase
+
+
+class MapEstimator:
+  r"""Mapping estimator between probability measures.
+
+  It estimates a map :math:`T` by minimizing the loss:
+  .. math:
+
+  \text{min}_{\theta}\; \Delta(T_\theta \sharp \mu, \theta)
+  + \lambda R(T_\theta)
+
+  where :math:`\Delta` is a fitting loss and :math:`R` is a regularizer.
+  :math:`\Delta` allows to fit the marginal constraint, i.e. transport
+  :math:`\mu` to  :math:`\nu` via :math:`T`, while :math:`R`
+  is a regularizer imposing an inductive bias on the learned map.
+
+  For instance, :math:`\Delta` can be the
+  {func}`~ott.tools.sinkhorn_divergence.sinkhorn_divergence`
+  and :math:`R` the {func}`~ott.solvers.nn.losses.monge_gap`
+  (see :cite:`uscidda:23`) for a given cost function :math:`c`.
+  In that case, it estimates a :math:`c`-OT map, i.e. a map :math:`T`
+  optimal for the Monge problem induced by :math:`c`.
+
+  Args:
+    dim_data: input dimensionality of data required for network init
+    model: network architecture for map :math:`T`.
+    optimizer: optimizer function for map :math:`T`.
+    fitting_loss: fitting loss :math:`Delta` to fit the marginal constraint
+    regularizer: regularizer :math:`R` to impose an inductive bias
+      on the map :math:`T`
+    num_train_iters: number of total training iterations
+    valid_freq: frequency with training and validation are logged
+    logging: option to return logs
+    rng: random key used for seeding for network initializations
+  """
+
+  def __init__(
+      self,
+      dim_data: int,
+      model: ModelBase,
+      optimizer: optax.OptState,
+      fitting_loss: Optional[Callable[[jnp.ndarray, jnp.ndarray],
+                                      float]] = None,
+      regularizer: Optional[Callable[[jnp.ndarray, jnp.ndarray], float]] = None,
+      lambd_regularizer: float = 1.,
```

Comment:
I'd prefer to call it `regularizer_strength`. Also it's missing description in the docstring.